JWBestPracticeApps/AdvancedPlayer/AdvancedPlayer/ViewController.swift

```diff
@@ -0,0 +1,299 @@
+// 4. Detect whether an upcoming ad is skippable
+// 5. Time observation for content
+// 6. Time observation for ads
+
+class ViewController: JWPlayerViewController,
+                      JWPlayerViewControllerDelegate,
+                      CustomControlsDelegate {
+
+    private let videoUrlString = "https://playertest.longtailvideo.com/adaptive/oceans/oceans.m3u8"
+    private let posterUrlString = "https://d3el35u4qe4frz.cloudfront.net/bkaovAYt-480.jpg"
+
+    // custom UI
+    var customControls: CustomControls!
+    var progressBarScrubbing: Bool = false
+    var skipOffset: Double?
+    var fullScreenVC: JWFullScreenViewController?
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        delegate = self
+
+        // Set the view's background color to black for better contrast.
+        view.backgroundColor = .black
+
+        // MARK: - 1. Hide & disable built-in controls/ad controls
+        interfaceBehavior = .hidden
+
+        // MARK: - 2. Add custom controls
+        customControls = UINib(nibName: "CustomControls", bundle: .main).instantiate(withOwner: nil, options: nil).first as? CustomControls
+        customControls.translatesAutoresizingMaskIntoConstraints = false
+        customControls.delegate = self
+        addCustomControls(toView: view)
+
+        // Set up the player.
+        setUpPlayer()
+    }
+
+    /**
+     Sets up the player with a simple configuration.
+     */
+    private func setUpPlayer() {
+        let videoUrl = URL(string:videoUrlString)!
+        let posterUrl = URL(string:posterUrlString)!
+
+        let playerItembuilder = JWPlayerItemBuilder()
+            .file(videoUrl)
+            .posterImage(posterUrl)
+        var playerItem: JWPlayerItem!
+        do {
+            playerItem = try playerItembuilder.build()
+        } catch {
+            // Handle player item build failure
+            print(error.localizedDescription)
+            return
+        }
+
+        let advertisingConfigBuilder = JWAdsAdvertisingConfigBuilder()
+            .vmapURL(URL(string: "https://s3.amazonaws.com/george.success.jwplayer.com/demos/vmap_midroll_preroll.xml")!)
+        var advertisingConfig: JWAdvertisingConfig!
+        do {
+            advertisingConfig = try advertisingConfigBuilder.build()
+        } catch {
+            // Handle advertising config build failure
+            print(error.localizedDescription)
+            return
+        }
+
+        let configBuilder = JWPlayerConfigurationBuilder()
+            .playlist([playerItem])
+            .advertising(advertisingConfig)
+        var config: JWPlayerConfiguration!
+        do {
+            config = try configBuilder.build()
+        } catch {
+            // Handle player item build failure
+            print(error.localizedDescription)
+            return
+        }
+
+        player.configurePlayer(with: config)
+    }
+
+    // MARK: - Add custom controls
+
+    func addCustomControls(toView view: UIView) {
```

Comment:
Some simple commenting in this method will help, to understand its purpose. The naming is fine, but we have to assume the external developers don't understand it.